### PR TITLE
server: check heartbeat activity.

### PR DESCRIPTION
**Pull request by disksing**
Discard inactive connections and periodically bind active connections. Used to circumvent problems that may occur by reconnecting.

### CODE REVIEW

File: server/grpc_service.go [hunk 1]
```diff
@@ -246,6 +250,25 @@ var errSendRegionHeartbeatTimeout = errors.New("send region heartbeat timeout")
...
 	stream pdpb.PD_RegionHeartbeatServer
 	closed int32
+	active int32
+}
+
+func newHeartbeatServer(stream pdpb.PD_RegionHeartbeatServer) *heartbeatServer {
+	s := &heartbeatServer{stream: stream}
+	go func() {
+		for {
+			time.Sleep(regionHeartbeatStreamMaxIdleTime)
```
Comment (siddontang) [inline server/grpc_service.go]: seem we can't quit this goroutine quickly, if the stream is connected and closed frequently, we may have many goroutines.

File: server/grpc_service.go [hunk 2]
```diff
@@ -275,6 +298,7 @@ func (s *heartbeatServer) Recv() (*pdpb.RegionHeartbeatRequest, error) {
 		atomic.StoreInt32(&s.closed, 1)
 		return nil, errors.Trace(err)
 	}
+	atomic.StoreInt32(&s.active, 1)
```
Comment (siddontang) [inline server/grpc_service.go]: How to set in-active?

**Comment by siddontang** [comment]
any test?

File: server/grpc_service.go [hunk 3]
```diff
@@ -246,6 +250,25 @@ var errSendRegionHeartbeatTimeout = errors.New("send region heartbeat timeout")
 type heartbeatServer struct {
 	stream pdpb.PD_RegionHeartbeatServer
 	closed int32
+	active int32
+}
+
+func newHeartbeatServer(stream pdpb.PD_RegionHeartbeatServer) *heartbeatServer {
```
Comment (Connor1996) [inline server/grpc_service.go]: seems it's called in nowhere

**Pull request closed without merging**